README.md
# telerivet-roster-api

Do something worthless to show how JIRA integrates.

Telerivet-focused bindings for the Roster API

These scripts are meant to be require()'d as a cloud script module in a Telerivet project to enable simple Roster API calls.  The module handles annoying tasks like:

* Transparently loading API endpoint and API key from a shared data table
* Persisting state (for example, client credentials) between different javascript runs in a Telerivet logic flow
* Wrapping HTTP errors into standard javascript errors, wrapping standard javascript errors into Telerivet `$variables`

## Quickstart

1.  First, this repository is added as as a ["Cloud Script Module"](https://telerivet.com/dashboard/a/add_script_module) with a given name, like: `ext/roster`

    Stable releases of the Roster API bindings are tagged with semantic versions, for example `v1.0.0`.  Generally production
    logic flows should import only stable releases as modules by specifying `vX.Y.Z` as the branch name.

    **CAVEAT**: For now, Telerivet cloud script modules must have GLOBALLY unique names. Thus, if your co-worker's account already added a module called `ext/roster`, you'll have to create your own name, like `ext/my-own-test-roster`. Sorry, we know that's weird.

1.  To use the Roster API in Telerivet "Logic Flows", add a "Run Custom Javascript" Action to the flow with the following lines:

    ```javascript
    var rosterAPI = require('ext/roster/api');
    catchAll(function() {

      // Do stuff with the roster here
      $result = rosterAPI.doStuff();

    }); // $error, $error_message set if errors are thrown
    ```

    For example, a call to get the client balance might look like:

    ```javascript
    var rosterAPI = require('ext/roster/api');
    catchAll(function() {

      $balance = rosterAPI.getClient($accountNumber).balance;

    });
    ```

1.  The Roster API requires a Roster endpoint (URL) and API key be specified - this can be done in the binding itself by:

    ```javascript
    var rosterAPI = require('ext/roster/api').attach("http://www.oaf.org/api/v0", "really-long-key-012345");
    ```

    ... but preferably can be done across-the-board in the project settings themselves by creating a new data table called "ExternalApis" in the project with three columns:

    | Name | URL | Key |
    | ------------- |:-------------:| -----:|
    | Roster | http://www.oaf.org/api/v0 | really-long-key-012345 |

    This way the Roster endpoint and key can be reconfigured without touching a lot of code inside the logic flows.

## Updates to the bindings

Semantic versions indicate breaking changes by bumping the major version number (`v1` to `v2`), and non-breaking feature additions by bumping the minor version number (`v2.2` to `v2.4`).

On Telerivet, the Roster API bindings are updated by changing to a new versioned branch in the Cloud Script Module for the bindings.  Minor-version updates *should* be safe and require no Telerivet logic flow changes... but a healthy level of paranoia is a great idea.  Major version updates generally require logic flow changes.

It's also possible to use multiple versions of the API bindings by registering differently named Cloud Script Modules - for example `ext/roster-v1.1` and `ext/roster-v2.0`.  This way new bindings can be tested and upgraded slowly in different logic flows one-by-one.

## API Bindings

### request(path, opts) => obj

```javascript
$result = rosterAPI.request("Clients", { params: { 'account' : '12345' } });
```

Raw API request functionality which uses the attached endpoint and API key of the `rosterAPI` to do a request at a certain URL path.  Returns the response content when successful, otherwise throws an HttpError containing the full response object and other data.

------

### authClient(accountNumber, [countryOrPhone], [accountPin]) => bool

```javascript
$isValidAndAuthorized = rosterAPI.authClient("12345");
```

Authorizes the client to access Roster, which (for now) simply requires a valid `accountNumber`, and stores the provided credentials for future calls (`accountPin` if provided).  The `countryOrPhone` parameter may be either the telerivet global `phone` object (default if not provided), a Roster country name, or an ISO country name.

`authClient` must be used before making other API calls, even without a PIN, otherwise `403 HttpErrors` will be thrown.  For now you can work around this with `rosterAPI.credentials = { key: rosterAPI.key }; rosterAPI.saveState();`, but this won't work once calls no longer include the account number.

### getClient(accountNumber, [countryOrPhone]) => obj

```javascript
$client = rosterAPI.getClient("12345");
```

Returns a client information object with at least the following:

```javascript
{
  GlobalClientId: "...",
  AccountNumber: (number),
  ClientName: "...", // full name
  FirstName: "...",
  LastName: "...",

  // active account details
  ClientId: (number),
  DistrictId: (number),
  DistrictName: "...",
  RegionId: (number),
  RegionName: "...",
  CountryId: (number),
  CountryName: "...",
  CreatedDate: (date),
  EnrollmentDate: (date),
  BannedDate: (date) | null,
  DeceasedDate: (date) | null,

  // aggregate account details
  EarliestCreatedDate: (date),
  EarliestEnrollmentDate: (date),
  LatestBannedDate: (date) | null,

  // current enrollment details (null if client has never enrolled with active account)
  GroupId: (number) | null,
  GroupName: "..." | null,
  SiteId: (number) | null,
  SiteName: "..." | null,

  // client accounts in reverse chronological order
  // Each account has a unique AccountGuid
  AccountHistory: [
    {
      AccountGuid: "...",
      ClientId: (number),
      AccountNumber: "...",
      DistrictId: (number),
      DistrictName: "...",
      RegionId: (number),
      RegionName: "...",
      CountryId: (number),
      CountryName: "..."
    },
    ...
  ],

  // balance history in reverse chronological order
  BalanceHistory: [
    {
      // enrollment details
      AccountGuid: "...",
      GroupId: (number),
      GroupName: "...",
      SiteId: (number),
      SiteName: "...",
      SeasonId: (number),
      SeasonName: "...",
      SeasonStart: "...",

      // repayment details
      TotalCredit: (number),
      TotalRepayment_IncludingOverpayments: (number),
      Balance: (number),
      CurrencyCode: "..."},
   ...
   ]
}
```

The `accountNumber` and `countryOrPhone` parameters are used similarly to the `authClient` call.

### isSerialNumberRegistered(productTypes, serialNum, accountNumber, [countryOrPhone]) => obj

```javascript
$result = rosterAPI.isSerialNumberRegistered("Sun King Home", "67890987", "12345");
```

Returns a result object with at least the following:

```javascript
{
  Result: "SerialNumberNotRegistered",
  SerialNumProduct: "Sun King Home"
}
```

The `productTypes` parameter may be either a single Roster product type (name of a Roster input) or an array of several product types.

The `accountNumber` and `countryOrPhone` parameters are used similarly to the `getClient` call.

The `Result` values are one of:

* `"SerialNumberNotRegistered"`
* `"SerialNumberIsRegisteredToAnotherClient"`
* `"SerialNumberIsAlreadyRegisteredToCurrentClient"`
* `"SerialNumberDoesNotBelongToPassedInputs"` - the serial number can't be found for these inputs in the Roster database
* `"SerialNumberAmbiguous"` - the serial number was found for two or more products of different types, so a result can't be returned.  This shouldn't happen if only one product type is specified (but rarely can due to an ongoing Roster data issue).

The `SerialNumProduct` is provided (if possible) to indicate the product type the serial number was found for - generally this is only useful if multiple product types are specified.

## Other Helpers

### [global].catchAll

```javascript
var rosterAPI = require('ext/roster/api');
catchAll(function() {

    $result = rosterAPI.makeBadCall();

}); // $error, $error_message set if errors are thrown

if ($error) {
   // Handle error case in javascript or in logic flow
}

```

The `catchAll` helper function wraps javascript errors in a helpful way so they can be used later in a Telerivet error-handling logic branch.  Because only global `$`-prefixed variables with `string|int` values are usable later in Telerivet logic flows, error handling isn't possible unless errors are reported in this form.

Certain error classes also report extra information (like `HttpError`'s `$error_status` and `$error_url`) by implementing a `toTelerivet()` method.

### parseAccountAndPin()

Helper to infer the account number, country, and PIN a user passes in from the raw content of a text message.

Account numbers are identified with a `#` prefix (`#12345678`), PINs are identified with `P` (`P123`), and country is inferred but may be specified by `@` (`@kenya`).

If `parsed.error != null` the parsing was unsuccessful and `parsed.error.message` may be returned to the user.  Alternately, `parsed.error.code` may be used to provide a custom message instead.


RELEASE_NOTES.md
# Release Notes

## Version 1.1.0
- Add more ISO countries to OAF mapper function
- README changes

## Version 1.0.2
- Bugfix: mock global 'phone', 'state', and 'content' variables in a recurring or manually triggered service
- Bugfix: successfully read ApiKey from Data Tables on first request

## Version 1.0.1
- Test versioning

## Version 1.0.0
- authClient() with optional PIN validation
- getClient()
- isSerialNumberRegistered() query-only call


test/test.js
// Tests the telerivet working stuff

// Setup underscore environment to match Telerivet
global._ = require('../node_modules/underscore');

// Explicitly hook up telerivet stuff when we need it
global.telerivetContext = {};

var assert = require('assert');

function Cursor(arr) {

    this.i = 0;
    this.arr = arr;

}

Cursor.prototype.hasNext = function() {
    return this.i < this.arr.length;
};

Cursor.prototype.next = function() {
    var value = this.arr[this.i];
    this.i++;
    return value;
};

function MockProject() {

    this.tables = {};
};

MockProject.prototype.queryDataTables = function(opts) {

    if (!('name' in opts)) return null;

    return new Cursor([this.tables[opts.name]]);
};

function MockTable(rows) {
    this.rows = rows;
    if (!rows) this.rows = [];
};

MockTable.prototype.queryRows = function(opts) {
    return new Cursor(this.rows);
};

// Mock telerivet object that gives custom reponses and logs
// requests
function MockRivet() {

    var responses = this.responses = [];

    this.httpClient = {
        request: function() {
            var response = responses.shift();
            return response;
        }
    };

    this.project = new MockProject();
    this.state = {
        vars: {}
    };
    this.phone = {};
}

var tests = [ // BEGIN TESTS

    function logNodeVersion() {
        console.log("Running with versions:");
        console.log(process.versions);
    },

    function testTrim() {

        // String trimming tests

        var utils = require('../utils');
        assert.equal("abcde", utils.trim(" abcde\n "));
        assert.equal("foo bar", utils.trim("foo bar "));
    },

    function testJoinURL() {

        // Test join logic for url roots and paths

        var utils = require('../utils');
        var joinURL = utils.joinURL;

        assert.equal(joinURL("http://www.google.com/", "query"),
            "http://www.google.com/query");
        assert.equal(joinURL("foo", "bar"), "foo/bar");
    },

    function testNestedGet() {

        // Nested path object get tests

        var utils = require('../utils');
        var nestedGet = utils.nestedGet;

        var obj = {
            a: {
                b: 1
            },
            c: {
                d: "ignore"
            },
            "c.d": 2,
            e: [{
                f: 3
            }, {
                f: 4
            }]
        };

        assert.equal(nestedGet("a.b", obj), 1);
        assert.equal(nestedGet("c.d", obj), 2);
        assert.equal(nestedGet("e.0.f", obj), 3);
        assert.equal(nestedGet("e.1.f", obj), 4);
    },

    function testFormat() {

        // Basic string format tests

        var utils = require('../utils');
        assert.equal("Hello World", utils.format("Hello {0}", ["World"]));
        assert.equal("Goodbye Moon 2", utils.format("Good{0} {1} {2}", ["bye", "Moon", 2]));
    },

    function testAccountParse() {

        // Tests parsing various account information from text messages

        var api = require('../api');

        var content = "#12345@kenya";

        var parsed = api.parseAccountAndPin(content);
        assert.equal(parsed.accountNumber, "12345");
        assert.equal(parsed.country, "kenya");
        assert(!parsed.accountPin);

        content = "@burundi#abcde P123";
        var parsed = api.parseAccountAndPin(content);
        assert.equal(parsed.accountNumber, "abcde");
        assert.equal(parsed.country, "burundi");
        assert.equal(parsed.accountPin, "123");

        content = "P1234 #12346 @zambia and more stuff";
        var parsed = api.parseAccountAndPin(content);
        assert.equal(parsed.accountNumber, "12346");
        assert.equal(parsed.country, "zambia");
        assert.equal(parsed.accountPin, "1234");
    },

    function testApiVersion() {

        // Tests super dumb version info coming back

        var api = require('../api');

        console.log("API Version", api.getVersion());

        assert(parseInt(api.getVersion().semver[0]) > 0);
    },

    function testPhoneContext() {

        // Tests parsing various account information from text messages

        var api = require('../api');

        var context = api.toPhoneContext("kenya");
        assert.equal(context.isoCountry, "KE");
        assert.equal(context.oafCountry, "Kenya");

        var context = api.toPhoneContext({
            country: "mw"
        });
        assert(context.phone.country, "mw");
        assert(context.isoCountry, "MW");
        assert(context.oafCountry, "Malawi");
    },

    function testAttachTable() {

        // Tests the API calls required to get client information

        var api = require('../api');

        api.telerivet = new MockRivet();
        var table = new MockTable([{
            vars: {
                name: "Roster",
                url: "www.roster.com",
                key: "foo"
            }
        }]);
        api.telerivet.project.tables["ExternalApis"] = table;

        api.dataTableAttach();
        assert.equal(api.url, "www.roster.com");
        assert.equal(api.key, "foo");
    },

    function testSaveRestore() {

        // Saving the API state between calls

        var api = require('../api');

        api.telerivet = new MockRivet();
        api.attach("URL", "APIKEY");

        var serialized = api.saveState();

        api.attach("URL2", "APIKEY2");

        assert.equal(api.url, "URL2");
        assert.equal(api.key, "APIKEY2");

        api.restoreState(serialized);

        assert.equal(api.url, "URL");
        assert.equal(api.key, "APIKEY");
    },

    function testAuthClient() {

        // Tests the API calls required to get client information

        var api = require('../api');

        api.attach("http://oaf.com", "APIKEY");
        api.telerivet = new MockRivet();

        api.telerivet.responses.push({
            status: 200,
            content: JSON.stringify({
                "Result": "ClientExists"
            })
        });
        var authd = api.authClient("CLIENTID", {
            country: "ke"
        }, "PIN");

        assert(authd);

        var request = api.requestLog[0];
        assert.equal(request[0], "http://oaf.com/sms/Validate");
        assert.equal(request[1].params.account, "CLIENTID");
        assert.equal(request[1].params.country, "Kenya");
        assert.equal(request[1].headers.Pin, "PIN");
        assert.equal(request[1].headers['X-OAF-Account'], "CLIENTID");
        assert.equal(request[1].headers.Authorization, "ApiKey APIKEY");
        assert.equal(api.credentials.accountCountry, "Kenya");
        assert.equal(api.credentials.accountNumber, "CLIENTID");
        assert.equal(api.credentials.accountPin, "PIN");
    },

    function testGetClient() {

        // Tests the API calls required to get client information

        var api = require('../api');

        api.attach("http://oaf.com", "APIKEY");
        api.telerivet = new MockRivet();
        api.telerivet.phone = {
            country: "KE"
        };

        // Auth with client pin
        api.telerivet.responses.push({
            status: 200,
            content: JSON.stringify({
                "Result": "ClientExists"
            })
        });
        assert(api.authClient("7890", null, "PIN"));

        // Check country encoded with accountNumber
        api.telerivet.responses.push({
            status: 200,
            content: JSON.stringify({
                "foo": "bar"
            })
        });
        api.requestLog = [];

        var client = api.getClient("7890");
        assert.equal(client.foo, "bar");

        var request = api.requestLog[0];
        assert.equal(request[0], "http://oaf.com/sms/Client");
        assert.equal(request[1].params.account, "7890");
        assert.equal(request[1].params.country, "Kenya");
        assert.equal(request[1].headers.Pin, "PIN");
        assert.equal(request[1].headers['X-OAF-Pin'], "PIN");
        assert.equal(request[1].headers['X-OAF-Account'], "7890");
        assert.equal(request[1].headers['X-OAF-Country'], "Kenya");
        assert.equal(request[1].headers.Authorization, "ApiKey APIKEY");
    },

    function testGetClientError() {

        // Tests errors thrown by API calls

        var api = require('../api');

        api.attach("http://oaf.com", "12345");
        api.telerivet = new MockRivet();

        // Check country encoded with accountNumber
        api.telerivet.responses.push({
            content: {
                message: "Server Error"
            },
            status: 500
        });

        try {
            var client = api.getClient("7890", "111");
            assert(false);
        } catch (err) {
            assert.equal(err.message, "Server Error");
            assert.equal(err.response.status, 500);
        }

        // Test telerivet wrapping
        api.telerivet.responses.push({
            content: JSON.stringify({
                message: "Server Error"
            }),
            status: 500
        });

        catchAll(function() {
            var client = api.getClient("7890", "111");
        });

        assert.equal($error, "HttpError");
        assert.equal($error_message, "Server Error");
        assert.equal($error_url, "http://oaf.com/sms/Client");
        assert.equal(JSON.parse($error_opts).params.account, "7890");
    },

    function testTrAssert() {

        var trassert = require('./trassert');

        trassert.ok(true);

        var err = null;
        try {
            trassert.ok(false);
        } catch (trErr) {
            err = trErr;
        }
        assert(err !== null);

        trassert.equal("a", "a");
        try {
            trassert.equal("a", "b");
        } catch (trErr) {
            err = trErr;
        }
        assert(err !== null);
    },

]; // END TESTS

for (var i = 0; i < tests.length; ++i) {
    console.log("---\nRunning test '" + tests[i].name + "'\n---");
    tests[i]();
}


test/trtest1.js
//
// Integration tests in a telerivet environment
//

var assert = require('./assert');

var api = require('../api');

var apiTable = project.getOrCreateDataTable("TestApis");
apiTable.delete();
apiTable = project.getOrCreateDataTable("TestApis");

apiTable.createRow({
    vars: {
        name: "Roster",
        url: "URL",
        key: "APIKEY",
    }
}).save();

api.dataTableAttach("TestApis");
assert.ok(state.vars[api.persistVar]);

assert.ok(false);

test/trtest2.js
//
// Integration tests in a telerivet environment
// (Second block)
//

var assert = require('./assert');

var api = require('../api');

assert.ok(state.vars[api.persistVar]);
assert.equal(api.url, "URL");
assert.equal(api.key, "APIKEY");

utils.js
//
// Helper functions required by all modules
// Since telerivet uses some weird custom javascript, we need to
// add some helpers here.
//

var trim = function(str) {
    return str.replace(/^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g, '');
};

var joinURL = function(a, b, sep) {
    if (a.charAt(a.length - 1) == '/')
        a = a.substr(0, a.length - 1);
    return a + "/" + b;
};

var nestedGet = function(key, obj) {

    if (key === '.') return obj;
    if (!_.isArray(obj) && !_.isObject(obj))
        return undefined;
    if (key in obj) return obj[key];

    var dotIndex = key.indexOf('.');
    if (dotIndex == -1) dotIndex = key.length;

    var dotKey = key.substring(0, dotIndex);
    var remKey = key.substring(dotIndex + 1);

    var dotValue = undefined;
    try {
        if (_.isArray(obj)) {
            dotKey = parseInt(dotKey);
        }
        var dotValue = obj[dotKey];
    } catch (ex) {
        return undefined;
    }

    if (remKey.length === 0) return dotValue;
    return nestedGet(remKey, dotValue);
};

// printf-style formatting: format("Hi {0}", "There") => "Hi There"
var format = function(str, args) {
    if (!args) args = [];
    if (!_.isArray(args))
        args = [args];

    return str.replace(/{(\d+)}/g, function(match, number) {
        return typeof args[number] != 'undefined' ?
            args[number] :
            match;
    });
};

var isoToOaf = {
    "KE": "Kenya",
    "UG": "Uganda",
    "ZM": "Zambia",
    "RW": "Rwanda",
    "TZ": "Tanzania",
    "BI": "Burundi",
    "ZM": "Zambia",
    "MM": "Myanmar",
    "MW": "Malawi",
    "NG": "Nigeria",
    "IN": "India",
    "ET": "Ethiopia"
};

var isoToOafCountry = function(isoCountry) {
    if (isoCountry in isoToOaf)
        return isoToOaf[isoCountry.toUpperCase()];
    // Passthrough to allow OAF names to just work too
    return isoCountry;
};

var oafToIsoCountry = function(oafCountry) {
    for (var iso in isoToOaf) {
        if (isoToOaf[iso].toUpperCase() == oafCountry.toUpperCase())
            return iso;
    }
    return oafCountry;
};

module.exports = {
    trim: trim,
    joinURL: joinURL,
    nestedGet: nestedGet,
    format: format,
    isoToOafCountry: isoToOafCountry,
    oafToIsoCountry: oafToIsoCountry
};


api.js
var utils = require('./utils');
var version = require('./version');

// Dump globals into a testable object we can
// use in modules.

if (!('telerivetContext' in global)) {

    global.telerivetContext = {
        // these variables are only defined within Telerivet's script engine in certain contexts (e.g. in response to a message)
        // they will be undefined in the context of a periodically-running service
        // https://telerivet.com/api/script#global

        phone: 'phone' in global ? phone : null,

        // telerivet's ContactServiceState -- we store the URL, apiKey, and credentials in this Object's customs vars
        state: 'state' in global ? state : {
            vars: {}
        },
        content: 'content' in global ? content : "",

        // these variables are always defined in Telerivet's script engine
        project: project,
        httpClient: httpClient
    };
}

//
// Start API we want to expose
//

function RosterAPI(telerivet) {

    this.telerivet = telerivet;

    this.url = null;
    this.key = null;
    this.credentials = null;

    this.requestLog = [];
    this.verbose = false;

    this.persistVar = '__RosterAPI__';
    this.restoreState();
};

RosterAPI.prototype.getVersion = function() {
    return version;
};

RosterAPI.prototype.restoreState = function(serialized) {

    if (!serialized) {
        if (!this.telerivet.state) return;
        if (!this.telerivet.state.vars[this.persistVar]) return;
        serialized = this.telerivet.state.vars[this.persistVar];
    }

    var state = JSON.parse(serialized);
    for (var k in state) {
        this[k] = state[k];
    }
};

RosterAPI.prototype.saveState = function() {

    if (!this.telerivet.state) return;

    var state = {
        url: null,
        key: null,
        credentials: null,
    };
    for (var k in state) {
        state[k] = this[k];
    }
    var serialized = JSON.stringify(state);
    this.telerivet.state.vars[this.persistVar] = serialized;
    return serialized;
};

RosterAPI.prototype.attach = function(url, key) {

    this.url = url;
    this.key = key;
    this.saveState();
    return this;
};

// We can attach to a URL/Key stored in the project "ExternalApis" data table
RosterAPI.prototype.dataTableAttach = function(tableName, project) {
    if (!tableName)
        tableName = "ExternalApis";

    if (!project)
        project = this.telerivet.project;

    var tableCur = project.queryDataTables({
        name: tableName
    });
    if (!tableCur.hasNext()) return false;
    var table = tableCur.next();

    var rowCur = table.queryRows();
    var row = null;
    while (rowCur.hasNext()) {
        var nextRowVars = rowCur.next().vars;
        if ('name' in nextRowVars && 'url' in nextRowVars && 'key' in nextRowVars && nextRowVars.name == "Roster") {
            row = nextRowVars;
            break;
        }
    }
    if (!row) return false;

    this.url = row.url;
    this.key = row.key;
    this.saveState();
    return true;
};

function HttpError(url, opts, response) {

    this.name = "HttpError";

    if (response.content &&
        (response.content.Message || response.content.message)) {
        this.message = response.content.Message || response.content.message;
    } else {
        this.message = JSON.stringify(response);
    }

    this.url = url;
    this.status = response.status;
    this.opts = opts;
    this.response = response;
}

HttpError.prototype = new Error();
HttpError.prototype.constructor = HttpError;
HttpError.prototype.toTelerivet = function() {
    return {
        $error_url: this.url,
        $error_status: this.status,
        $error_opts: JSON.stringify(this.opts),
        $error_response: JSON.stringify(this.response)
    };
};

RosterAPI.HttpError = HttpError;

RosterAPI.prototype.request = function(path, opts) {

    if (!this.url)
        this.dataTableAttach();

    if (!('headers' in opts)) opts.headers = {};
    var credentials = opts.credentials || this.credentials;

    if (credentials) {
        opts.headers['Authorization'] = "ApiKey " + this.key;
        opts.headers['X-OAF-Account'] = credentials.accountNumber;
        opts.headers['X-OAF-Country'] = credentials.accountCountry;
        var accountPin = credentials.accountPin ? credentials.accountPin : "";
        opts.headers['Pin'] = accountPin;
        opts.headers['X-OAF-Pin'] = accountPin;
    }

    if (!('Accept' in opts.headers))
        opts.headers['Accept'] = 'application/json';

    var fullURL = utils.joinURL(this.url, path);

    this.requestLog.push([fullURL, opts]);
    this.saveState();

    if (this.verbose) {
        console.log("Requesting:\n  " + fullURL + "\n options:\n  " + JSON.stringify(opts));
    }

    if (this.verbose) {
        console.log("Requesting:\n  " + fullURL + "\n options:\n  " + JSON.stringify(opts));
    }

    var response = this.telerivet.httpClient.request(fullURL, opts);

    // JSONify content if required
    if (response.content) {
        var contentType = response['Content-Type'];
        if (contentType && contentType.indexOf('application/json') === 0) {
            response.content = JSON.parse(response.content);
        } else if (!contentType && opts.headers['Accept'].indexOf('application/json') === 0) {
            try {
                response.content = JSON.parse(response.content);
            } catch (err) {
                // fail to parse
            }
        }
    }

    if (this.verbose) {
        console.log("Response:\n  " + JSON.stringify(response));
    }

    // 200s - ok
    // 300s - redirects
    // 400s/500s - errors
    if (response.status >= 300)
        throw new HttpError(fullURL, opts, response);

    return response.content;
};

var parseAccountNum = function(content) {
    // #12345 other stuff, #12345@kenya
    var match = content.match(/\#([^@\s]+)/);
    if (!match) return null;
    return match[1];
};

var parseCountry = function(content) {
    // @kenya other stuff
    var match = content.match(/\@(\w+)/);
    if (!match) return null;
    return match[1];
};

var parsePin = function(content) {
    // P1234 other stuff
    var match = content.match(/P([0-9]+)/);
    if (!match) return null;
    return match[1];
};

RosterAPI.prototype.parseAccountAndPin = function(content) {

    var parsed = {};

    parsed.accountNumber = parseAccountNum(content);
    parsed.accountPin = parsePin(content);
    parsed.country = parseCountry(content);

    return parsed;
};

RosterAPI.prototype.toPhoneContext = function(countryOrPhone) {

    if (countryOrPhone == null) {
        // in a scheduled service, the 'phone' will be null, so users must specify a country
        if (this.telerivet.phone == null)
            throw new Error("Please specify a country");
        else
            countryOrPhone = this.telerivet.phone;
    }

    var phoneContext = {};

    if (_.isString(countryOrPhone)) {
        phoneContext.phone = null;
        phoneContext.isoCountry = countryOrPhone.toUpperCase();
    } else {
        phoneContext.phone = countryOrPhone;
        phoneContext.isoCountry = phoneContext.phone.country.toUpperCase();
    }

    phoneContext.isoCountry =
        utils.oafToIsoCountry(phoneContext.isoCountry);
    phoneContext.oafCountry =
        utils.isoToOafCountry(phoneContext.isoCountry);

    return phoneContext;
};

RosterAPI.prototype.authClient = function(accountNumber, countryOrPhone, accountPin) {

    var phoneContext = this.toPhoneContext(countryOrPhone);

    var credentials = {
        accountNumber: accountNumber,
        accountCountry: phoneContext.oafCountry,
        accountPin: accountPin
    };

    var path = utils.format('sms/Validate', []);

    var opts = {
        method: 'GET',
        params: {
            account: accountNumber,
            country: phoneContext.oafCountry
        },
        headers: {
            'Accept': 'application/json'
        },
        credentials: credentials
    };

    var content = null;
    try {
        content = this.request(path, opts);
    } catch (err) {

        if (!(err instanceof HttpError))
            throw err;

        if (!(err.status === 403))
            throw err;

        // Authenticating is kind of weird right now, since we don't have
        // a dedicated endpoint for it.  Use Validate as a test for now.
        content = {
            Result: 'ClientDoesNotExist'
        };
    }

    if (content.Result == 'ClientExists') {
        this.credentials = credentials;
        this.saveState();
        return true;
    }

    return false;
};

RosterAPI.prototype.getClient = function(accountNumber, countryOrPhone) {

    var phoneContext = this.toPhoneContext(countryOrPhone);

    var path = utils.format('sms/Client', []);

    var opts = {
        method: 'GET',
        params: {
            account: accountNumber,
            country: phoneContext.oafCountry
        },
        headers: {
            'Accept': 'application/json'
        }
    };

    return this.request(path, opts);
};

RosterAPI.prototype.isSerialNumberRegistered = function(
    productTypes, serialNum, accountNumber, countryOrPhone) {

    var phoneContext = this.toPhoneContext(countryOrPhone);

    if (!_.isArray(productTypes))
        productTypes = [productTypes];

    var path = utils.format('sms/IsSerialNumberRegistered', []);

    var opts = {
        method: 'GET',
        params: {
            account: accountNumber,
            country: phoneContext.oafCountry,
            inputNames: productTypes,
            serialNum: serialNum
        },
        headers: {
            'Accept': 'application/json'
        }
    };

    return this.request(path, opts);
};

global.catchAll = function(todo) {

    try {
        todo();
        global.$error = "";
    } catch (error) {

        global.$error = error.name;
        global.$error_message = error.message;

        if (error.toTelerivet) {
            var trError = error.toTelerivet();
            for (var k in trError) {
                global[k] = trError[k];
            }
        }
    }
};

//
// Exports
//

var rosterAPI = new RosterAPI(telerivetContext);

module.exports = rosterAPI;
